A1-http-connect-proxy-support.md

```diff
@@ -1,222 +1,242 @@
-resolution and per-call load balancing internally to the backend servers.
-The trade-off here is that this approach requires that the gRPC-level
-proxy be trusted by all internal servers, which has security implications
-(e.g., a successful attack on the proxy would yield priviledged access
-on all internal servers).
+mapper and the name resolver both have to agree on the list of known proxy
+addresses. So, for example, if the proxy mapper implementation is getting the
+list of proxy addresses from a local file, the service owner would need to first
+push an updated list that contains the new proxy address to all clients. Then,
+once all clients have been updated, the new proxy can be added to DNS.
+
+Note that another alternative for case 3 would be to use a gRPC-level proxy
+instead of a TCP-level proxy, where the proxy natively speaks gRPC. The idea is
+that the proxy accepts a gRPC request from the external client and then uses its
+own client on the internal side, which does the name resolution and per-call
+load balancing internally to the backend servers. The trade-off here is that
+this approach requires that the gRPC-level proxy be trusted by all internal
+servers, which has security implications (e.g., a successful attack on the proxy
+would yield priviledged access on all internal servers).
 
 ## Implementation
 
-In C-core, the HTTP CONNECT handshaker was already implemented.
-The following additional changes have also been made:
+### C-Core
+
+In C-core, the HTTP CONNECT handshaker was already implemented. The following
+additional changes have also been made:
 
-- [grpc/grpc#9383](https://github.com/grpc/grpc/pull/9383) (changes
-  HTTP CONNECT handshaker to be triggered via channel args)
-- [grpc/grpc#9372](https://github.com/grpc/grpc/pull/9372) (adds proxy
-  mapper hook for address rewriting)
-- [grpc/grpc#9557](https://github.com/grpc/grpc/pull/9557) (adds proxy
-  mapper hook for name rewriting)
+-   [grpc/grpc#9383](https://github.com/grpc/grpc/pull/9383) (changes HTTP
+    CONNECT handshaker to be triggered via channel args)
+-   [grpc/grpc#9372](https://github.com/grpc/grpc/pull/9372) (adds proxy mapper
+    hook for address rewriting)
+-   [grpc/grpc#9557](https://github.com/grpc/grpc/pull/9557) (adds proxy mapper
+    hook for name rewriting)
+
+C-Core checks the following places to determine the HTTP proxy to use, stopping
+at the first one that is set:
+
+1.  `GRPC_ARG_HTTP_PROXY` channel arg
+2.  `grpc_proxy environment` variable
+3.  `https_proxy` environment variable
+4.  `http_proxy` environment variable
+
+If none of the above are set, then no HTTP proxy will be used.
+
+If an HTTP proxy to be used is found, C-Core then checks the follows places to
+exclude traffic destined to listed hosts from going through the proxy determined
+above, again stopping at the first one that is set:
+
+1.  `no_grpc_proxy` environment variable
+2.  `no_proxy`environment variable
```

Comment:
Can we document how the hosts are listed in these environment variables? Comma separated list? IP address only? Names? Can names and IP addresses be mixed? How are names resolved to IP addresses (or they are not?) Where is the matching of listed hosts to actual gRPC connection attempt being done? If the `no_proxy` list includes a name `foobar` which resolves to `10.1.2.3` and for an xDS channel if there is an endpoint `10.1.2.3` will it be properly excluded? The other way? e.g. if the exclude list contains IP address `10.1.2.3` and the caller uses name `foobar` to connect and `foobar` resolves to multiple addresses (through DNS) one of which happens to be `10.1.2.3` will it only exclude the subchannel to that IP address?

Reply:
Added details on how this works. https://github.com/grpc/grpc/pull/31119 improves the working of `no_proxy` but that is not part of a release yet.